DAS/das/utils/logger.py

```diff
@@ -0,0 +1,235 @@
+    broker_address : str
+        The IP address that the MQTT broker lives on
+    verbose : bool
+        Specifies whether the outgoing MQTT data and warnings are printed
+
+    Attributes
+    ----------
+    _VERBOSE: bool
+        Specifies whether the incoming MQTT data and warnings are printed
+    _BROKER_ADDRESS: bool
+        The IP address of the MQTT broker
+    _log_data: list(dict)
+        A list of all of the rows in the log file stored in dict format
+    """
+
+    def __init__(self, filepath: str, broker_address: str = "localhost", verbose: bool = True) -> None:
+        self._VERBOSE = verbose
+        self._BROKER_ADDRESS = broker_address
+
+        # Read in data from log file and save each row in _log_data list
+        log_file = open(filepath, "r")
+        csv_reader = csv.DictReader(
+            log_file,
+            delimiter=CsvConfig["delimiter"],
+            quotechar=CsvConfig["quotechar"],
+            quoting=CsvConfig["quoting"],
+            skipinitialspace=CsvConfig["skipinitialspace"])
+
+        self._log_data = []
+        for row in csv_reader:
+            # Convert time delta to float
+            row["time_delta"] = float(row["time_delta"])
+            self._log_data.append(row)
+
+    def play(self, speed: float = 1) -> None:
+        """Play the logged data at a certain speed using an async function.
+
+        Parameters
+        ----------
+        speed : float
+            Speed multiplier to determine how fast to send out the data. A higher value means faster.
+        """
+
+        if self._VERBOSE:
+            print(f"⚡ Playback initiated at {speed}x speed ⚡")
+
+        # Run the event loop to issue out all of the MQTT publishes
+        asyncio.run(self._publish(speed))
+
+    async def _publish(self, speed):
+        """Async function that collects all the necessary publish functions and gathers them to then be run by the
+        event loop.
+
+        Parameters
+        ----------
+        speed : float
+            Speed multiplier to determine how fast to send out the data. A higher value means faster.
+        """
+
+        async def _publish_aux(row):
+            scaled_sleep = row["time_delta"]/speed
+            await asyncio.sleep(scaled_sleep)
+
+            if self._VERBOSE:
+                print(
+                    f"{round(row['time_delta'], 5): <10} | {round(scaled_sleep, 5): <10} | {row['mqtt_topic']: <50} | {row['message']}")
+
+            try:
+                publish.single(
+                    row["mqtt_topic"],
+                    row["message"],
+                    hostname=self._BROKER_ADDRESS)
+            except Exception as e:
+                print(f"ERROR: {e}")
+
+        publish_queue = []
+        for row in self._log_data:
+            publish_queue.append(_publish_aux(row))
+
+        # Load all async operation at the start using gather
+        await asyncio.gather(*publish_queue, return_exceptions=True)
```

Comment:
I originally implemented it without async using `time.sleep`. I expected there would be a small difference between playback and actual time. Turns out there was a difference of 1s every minute. So for an hour recording we would be about 1m behind if we played it back.

The implementation in async is also much easier to understand and requires less math.

Reply:
It could be implemented fairly easily with `time.sleep`, just by storing the time which the last message was sent and then calculating how long until the time the next message should be sent. I think my main concern with async is that spawning thousands of threads feels messy and could effect performance (maybe, unsure) and feels unnecessary when the same logic could be achieved with a loop that people newer to coding would understand.